### ISSUE TITLE
List expression not allowed in type annotation   Use List[T] instead.

### ISSUE THREAD

**Issue by BigDream324** [microsoft/pyright#1321]
**Describe the bug**
when i define a function with multi type parameter, it report an error in editor.
function like this:
```python
    def _format_datetime(self, datetime_d: [str, pd.Timestamp]):
        datetime_d = pd.Timestamp(datetime_d)
        return datetime_d.strftime(self.calendar_format)
```
editor show error like this:
![image](https://user-images.githubusercontent.com/4465571/103183548-1bc14780-48ee-11eb-8bdc-8e46a752b7aa.png)

**Expected behavior**
the code runs well, so it should not report an error here.

**VS Code extension or command-line**
VS Code version:
![image](https://user-images.githubusercontent.com/4465571/103183588-5f1bb600-48ee-11eb-8b9b-512b5ce62de6.png)

**Additional context**

I write code in a container enviorment and I don't know whether the issue is related to this.

**Comment by shamrin** [comment]
I think Pyright is correct here. You have to use [Union][1] for a something that could be one of several types:

```python
from typing import Union
def _format_datetime(sef, datetime_d: Union[str, pd.Timestamp]):
    ...
```

[1]: https://docs.python.org/3/library/typing.html#typing.Union

(It's not surprising that your code runs well. Types has no effect on program execution in Python. They are only for type checker like Pyright.)

@erictraut, should Pyright mention Union in the error message as well? For example:

> List expression not allowed in type annotation
>   Use List[T] or Union instead

**Comment by erictraut** [comment]
You are using an incorrect syntax for your type annotation. As @shamrin said, you need to use an expression of the form `Union[x, y, z]` for union types. If you'e using the alpha (pre-release version) of Python 3.10, you can also make use of the alternate union syntax which takes the form `x | y | z`.

**Comment by erictraut** [comment]
I've also updated the error message text to suggest `Union[T1, T2]` in addition to `List[T]`.

**Comment by joep-source** [comment]
@erictraut do you know if it's possible to suppress this error? Thanks in advance!

**Comment by JelleZijlstra** [comment]
I don't maintain pyright but I think it's unlikely that they'll add support for disabling this error, since the type annotation is clearly invalid in terms of the Python type system and it's not obvious what it means.

Personally I would interpret that annotation as meaning "a two-element list containg a str and a Timetamp", so suggesting Union as an alternative may even be misleading.

**Comment by erictraut** [comment]
This is an invalid type annotation. There's no way to suppress the error short of a `# type: ignore` comment. I wouldn't recommend that approach though. If you want to enable static type checking, you should use only legal syntax and semantics for type annotations. Is there a reason why you can't simply fix the type annotation so it's legal?

**Comment by joep-source** [comment]
@erictraut thanks for replying! I'll have to stick with this illegal type annotation because it's all over the codebase and changing it won't be accepted for a PR. We're working on Python 3.7, in Python 3.9 becomes this type annotation legal.

**Comment by erictraut** [comment]
This syntax is not a legal type annotation in Python 3.9 either. You will need to fix this in your code base if you want to use static type checking.

In Python 3.10, you can use the `|` operator to indicate a union (e.g. `A | B`). Prior to Python 3.10, you should use `Union[A, B]` to indicate a union.

**Comment by trolliama** [comment]
I had the same error, but it wasn't fixed with the Union class

![image](https://user-images.githubusercontent.com/21108976/122595540-04624c00-d03f-11eb-8f62-403df6fa1b8b.png)

I added the Union statement but this give me another error

![image](https://user-images.githubusercontent.com/21108976/122595601-17751c00-d03f-11eb-9d32-218bc23af81c.png)

**Comment by BradDorney** [comment]
@erictraut Can `typing.no_type_check`/`no_type_check_decorator` be made to suppress the error? That is a way of explicitly marking that annotations are not type hints, so static type checkers ignore them, and makes it perfectly legal to use "non-type" annotations.

**Comment by erictraut** [comment]
@BradDorney, pyright doesn't support @no_type_check, and there are no plans to support it.

The Python static type system has evolved substantially since the early days of PEP 484. Its syntactical and semantic rules are now well established.

If you want to use non-standard "non-type" annotations, you can employ `Annotated`, introduced in [PEP 593](https://peps.python.org/pep-0593/).
